monai/networks/blocks/crf.py

```diff
@@ -0,0 +1,140 @@
+        bilateral_spatial_sigma: float = 5.0,
+        bilateral_color_sigma: float = 0.5,
+        gaussian_spatial_sigma: float = 5.0,
+        update_factor: float = 3.0,
+        compatability_kernel_range: int = 1,
+        iterations: int = 5,
+    ):
+        """
+        Args:
+            bilateral_weight: the weighting of the bilateral term in the message passing step.
+            gaussian_weight: the weighting of the gaussian term in the message passing step.
+            bilateral_spatial_sigma: standard deviation in spatial coordinates for the bilateral term.
+            bilateral_color_sigma: standard deviation in color space for the bilateral term.
+            gaussian_spatial_sigma: standard deviation in spatial coordinates for the gaussian term.
+            update_factor: determines the magnitude of each update.
+            compatability_kernel_range: the range of the kernel used in the compatability convolution.
+            iterations: the number of iterations.
+        """
+        super(CRF, self).__init__()
+        self.bilateral_weight = bilateral_weight
+        self.gaussian_weight = gaussian_weight
+        self.bilateral_spatial_sigma = bilateral_spatial_sigma
+        self.bilateral_color_sigma = bilateral_color_sigma
+        self.gaussian_spatial_sigma = gaussian_spatial_sigma
+        self.update_factor = update_factor
+        self.compatability_kernel_range = compatability_kernel_range
+        self.iterations = iterations
+
+    def forward(self, input_tensor: torch.Tensor, reference_tensor: torch.Tensor):
+        """
+        Args:
+            input_tensor: tensor containing initial class logits.
+            referenece_tensor: the reference tensor used to guide the message passing.
+
+        Returns:
+            output (torch.Tensor): output tensor.
+        """
+
+        # useful values
+        spatial_dim = input_tensor.dim() - 2
+        class_count = input_tensor.size(1)
+        padding = self.compatability_kernel_range
+
+        # constructing spatial feature tensor
+        spatial_features = _create_coordinate_tensor(reference_tensor)
+
+        # constructing final feature tensors for bilateral and gaussian kernel
+        bilateral_features = torch.cat(
+            [spatial_features / self.bilateral_spatial_sigma, reference_tensor / self.bilateral_color_sigma], dim=1
+        )
+        gaussian_features = spatial_features / self.gaussian_spatial_sigma
+
+        # compatability matrix (potts model (1 - diag) for now)
+        compatability_matrix = _potts_model_weights(class_count).to(device=input_tensor.device)
+
+        # expanding matrix to kernel
+        compatability_kernel = _expand_matrix_to_kernel(
+            compatability_matrix, spatial_dim, self.compatability_kernel_range
+        )
+
+        # choosing convolution function
+        conv = [conv1d, conv2d, conv3d][spatial_dim - 1]
+
+        # seting up output tensor
+        output_tensor = softmax(input_tensor, dim=1)
```

Comment:
just for clarification... I remember there were some normalisation factors https://github.com/SamuelJoutard/Permutohedral_attention_module/blob/master/CRF/crf.py#L103, probably from https://github.com/cvlab-epfl/densecrf/blob/c16e3c50d214ab8b4727c0d471f888240020d757/libDenseCRF/densecrf.cpp#L421, but couldn't recall the details... are they implemented here?

Reply:
This was a conscious choice. In my implementation I found them to have little effect on the output. But worth noting as a divergence from previous implementations.